src/main/scala/org/broadinstitute/dsde/workbench/sam/util/ShadowRunner.scala

```diff
@@ -0,0 +1,44 @@
+package org.broadinstitute.dsde.workbench.sam.util
+import cats.effect._
+
+import scala.concurrent.duration.MILLISECONDS
+import scala.concurrent.ExecutionContext
+
+trait ShadowRunner {
+  implicit val executionContext: ExecutionContext
+  val clock: Clock[IO]
+
+  protected def runWithShadow[T](functionName: String, real: IO[T], shadow: IO[T]): IO[T] = {
+    for {
+      realTimedResult <- measure(real)
+      _ <- measure(shadow).runAsync {
+        case Left(regrets) =>
+          reportShadowFailure(functionName, regrets)
+          IO.unit
+        case Right(shadowTimedResult) =>
+          reportResult(functionName, realTimedResult, shadowTimedResult)
+          IO.unit
+      }.toIO
+    } yield {
+      realTimedResult.result
+    }
+  }
+
+  protected def reportShadowFailure(functionName: String, regrets: Throwable): Unit = {
```

Comment:
I would make these `IO[Unit]` just so that you don't need to worry about when exactly this eager function will be evaluated.